pstoremem/addr_book.go

```diff
@@ -134,6 +154,48 @@ func (mab *memoryAddrBook) AddAddr(p peer.ID, addr ma.Multiaddr, ttl time.Durati
 // (time-to-live), after which the address is no longer valid.
 // This function never reduces the TTL or expiration of an address.
 func (mab *memoryAddrBook) AddAddrs(p peer.ID, addrs []ma.Multiaddr, ttl time.Duration) {
+	// if we have a valid peer record, ignore unsigned addrs
+	peerRec := mab.GetPeerRecord(p)
+	if peerRec != nil {
+		return
```

Comment:
This is a bit of a tricky case. We _don't_ want to add random addresses from the network, but we _might_ want to add our own observations (not to advertise, just for local use).

Reply:
Concrete problem: We have an _old_ record for a peer and now ask IPFS to dial the peer at a new address. IPFS will _try_ to add the new address to the peerstore and _fail_.

Reply:
Yeah. We might have to bite the bullet and extend the write methods (Add, Set) to accept option flags, or provenances. This really steps on the toes of the proposal for peerstore v2, though.

Reply:
Ideally, I'd like to have different "levels" of addresses and, mark the address "level" when returning it (locally set > signed > hearsay). However, that might be a bit tricky to work into the API.